lecture-notes/3-bld.md

```diff
@@ -323,3 +328,156 @@ Make has been around a long time and its performance and correctness are
+
+In the below example, we have a simply expanded variable `EXE` that can be used
+throughout the Makefile to refer to the desired executable name. This allows
+for a single point of truth.
+
+```make
+EXE := main
+$(EXE): main.o log.o
+	gcc main.o log.o -o $(EXE)
+main.o: main.c log.h
+	gcc -c main.c
+log.o: log.c log.h
+	gcc -c log.c
+.PHONY: clean
+clean:
+	rm -rf *.o $(EXE)
+```
+
+(While variables can normally have any name, the convention is to use
+upper-case characters.)
+
+**Recursively expanded** means that variable references and function calls on
+the right hand side happen when the variable is *used* (in the *target-update*
+phase for recipe usage). This structure allows complex templates to be stored
+in a variable and used with many different values.
+
+Expansions of both kinds of variables can occur anywhere, and use `$(VAR)`
+syntax.
+
+In the below example, we have a mix of simply and recursively expanded
+variables. `LOG_TARGET` is recursively expanded (note the `=`) and `CFLAGS` is
+simply expanded (note the `:=`).
+
+```make
+LOG_TARGET = echo "Current target: $@" ; echo "Prerequisites: $^"
+CFLAGS := -Wall
+CFLAGS := $(CFLAGS) -Werror
+main: main.o log.o
+	$(LOG_TARGET)
+	gcc $(CFLAGS) $^ -o $@
+main.o: main.c log.h
+	$(LOG_TARGET)
+	gcc $(CFLAGS) -c $< -o $@
+log.o: log.c log.h
+	$(LOG_TARGET)
+	gcc $(CFLAGS) -c $< -o $@
+```
+
+Because `LOG_TARGET` is recursively expanded, it is expanded at each use site
+in the Makefile: in `main`, in `main.o`, and in `log.o`. Every time it is
+expanded, the variable references inside it see the automatic variables (those
+funky-looking variables) local to that rule.
+
+Speaking of automatic variables...
+
+### Automatic variables
+
+Similar to how C gives you `__FILE__` and `__LINE__` and `__func__`, Make also
+defines some variables for you within rules. These special variables are called
+*automatic variables* and help refer to different parts of the rule so you, the
+programmer, don't repeat yourself too much.
+
+There are many automatic variables, particuarly in the GNU implementation of
+Make, but you will probably see these three the most: `$@`, `$^`, and `$<`. In
+order:
+
+* `$@` refers to the name of the target. In our example above, this is used in
+  the `gcc` invocations to provide an output filename for the `-o` option. In
+  the `main` rule, `$@` expands to `main`, in the `main.o` rule it expands to
+  `main.o`, etc.
+* `$^` refers to the names of all of the prerequisites, separated by spaces. In
+  the `main` rule, for example, it refers to `main.o log.o`. It is used here to
+  provide the names of all of the object files to GCC so they can be linked
+  together. It is *not* used in the `.o` rules because generally you do not
+  pass `.h` files to the compiler (they are `include`d).
+* `$<` <!-- TODO -->
+
+### Simple vs recursive expansion
+### Functions
+
+### Make variables and functions are textual
```

Comment:
Should definitely talk about how quoting differs/interacts between Make and shell in this section.